File: app/src/main/java/com/jetsetslow/omniterm/data/ssh/SshHostKeyTrust.kt
package com.jetsetslow.omniterm.data.ssh

import android.content.Context
import android.content.SharedPreferences
import android.util.Base64
import com.jcraft.jsch.HostKey
import com.jcraft.jsch.HostKeyRepository
import com.jcraft.jsch.UserInfo
import com.jetsetslow.omniterm.data.KnownHost
import com.jetsetslow.omniterm.data.SecretStore
import kotlinx.coroutines.CompletableDeferred
import kotlinx.coroutines.runBlocking
import kotlinx.coroutines.withTimeoutOrNull

data class HostKeyApprovalRequest(
    val host: String,
    val keyType: String,
    val fingerprint: String,
    val deferred: CompletableDeferred<Boolean>,
)

internal object SshHostKeyTrust {
    private const val PREFS = "ssh_host_key_trust"
    private const val APPROVAL_TIMEOUT_MS = 120_000L

    @Volatile private var prefs: SharedPreferences? = null

    /**
     * The active AppViewModel's first-connect approval handler. Registration is owner-token based:
     * a retiring ViewModel may clear only its own handler, never a newer instance's handler.
     */
    private data class ApprovalRegistration(
        val owner: Any,
        val handler: (HostKeyApprovalRequest) -> Unit,
    )

    @Volatile private var approvalRegistration: ApprovalRegistration? = null
    private val firstPinCommitLock = Any()

    @Synchronized
    fun registerApprovalHandler(owner: Any, handler: (HostKeyApprovalRequest) -> Unit) {
        approvalRegistration = ApprovalRegistration(owner, handler)
    }

    @Synchronized
    fun clearApprovalHandler(owner: Any) {
        if (approvalRegistration?.owner === owner) approvalRegistration = null
    }

    fun init(context: Context) {
        prefs = context.applicationContext.getSharedPreferences(PREFS, Context.MODE_PRIVATE)
    }

    fun repository(): HostKeyRepository = TofuHostKeyRepository(
        prefs ?: error("SshHostKeyTrust.init must be called before opening SSH sessions"),
    )

    fun listKnownHosts(): List<KnownHost> {
        val p = prefs ?: return emptyList()
        return p.all.mapNotNull { (storageKey, rawValue) ->
            val decoded = readStoredStatic(rawValue as? String) ?: return@mapNotNull null
            val parts = storageKey.split("|", limit = 2)
            val host = parts.getOrNull(0)?.takeIf { it.isNotBlank() } ?: return@mapNotNull null
            val keyType = parts.getOrNull(1) ?: ""
            val fp = runCatching {
                val bytes = android.util.Base64.decode(decoded, android.util.Base64.NO_WRAP)
                val md = java.security.MessageDigest.getInstance("SHA-256")
                android.util.Base64.encodeToString(md.digest(bytes), android.util.Base64.NO_WRAP)
                    .trimEnd('=').let { "SHA256:$it" }
            }.getOrDefault("(unknown)")
            KnownHost(host, keyType, fp)
        }
    }

    private fun readStoredStatic(raw: String?): String? {
        if (raw.isNullOrBlank()) return null
        return if (SecretStore.isEncrypted(raw)) SecretStore.decrypt(raw) else raw
    }

    /** All pinned host keys as storage key ("host|type") → base64 public key, decrypted —
     *  used by the app backup so a restored fleet keeps its trust store. */
    fun exportEntries(): Map<String, String> {
        val p = prefs ?: return emptyMap()
        return p.all.entries.mapNotNull { (key, value) ->
            readStoredStatic(value as? String)?.let { key to it }
        }.toMap()
    }

    /** Import pinned host keys from a backup (encrypting at rest). Existing pins always win —
     *  a backup must never silently replace a key this device has already verified. */
    fun importEntries(entries: Map<String, String>) {
        val p = prefs ?: return
        val edit = p.edit()
        var changed = false
        for ((key, encoded) in entries) {
            if (!key.contains('|') || encoded.isBlank() || p.contains(key)) continue
            edit.putString(key, SecretStore.encrypt(encoded) ?: encoded)
            changed = true
        }
        if (changed) edit.apply()
    }

    /** Replace the complete trust snapshot, used only to compensate a failed transactional restore. */
    fun replaceEntries(entries: Map<String, String>) {
        val p = prefs ?: return
        val edit = p.edit().clear()
        entries.forEach { (key, encoded) ->
            if (key.contains('|') && encoded.isNotBlank()) {
                edit.putString(key, SecretStore.encrypt(encoded) ?: encoded)
            }
        }
        check(edit.commit()) { "Could not restore SSH host-key snapshot" }
    }

    /**
     * Every storage-key alias prefix a given host/port can be pinned under. Matches the aliasing
     * used by [hasPinnedKey] and [removeHost] so callers don't re-derive the convention.
     */
    private fun storageAliases(host: String, port: Int): Set<String> =
        setOf(host, "$host:$port", "[$host]:$port")

    /**
     * Filter a backup's exported trust entries ("<alias>|<type>" → key) down to those whose host
     * belongs to one of [hosts] (host to port). Used at restore time so pinned keys are imported
     * only for servers actually restored — orphaned trust entries for skipped/limited hosts are
     * dropped. Returns the kept entries; the count of dropped entries is `entries.size - result.size`.
     */
    fun filterEntriesForHosts(entries: Map<String, String>, hosts: Collection<Pair<String, Int>>): Map<String, String> {
        if (entries.isEmpty()) return entries
        val aliases = hosts.flatMap { (host, port) -> storageAliases(host, port) }.toSet()
        return entries.filter { (key, _) ->
            aliases.any { alias -> key.startsWith("$alias|") }
        }
    }

    /** True when at least one host key (any type) is already pinned for this host. */
    fun hasPinnedKey(host: String, port: Int): Boolean {
        val trustedPrefs = prefs ?: return false
        val aliases = storageAliases(host, port)
        return trustedPrefs.all.keys.any { key -> aliases.any { alias -> key.startsWith("$alias|") } }
    }

    fun removeHost(host: String, port: Int) {
        val trustedPrefs = prefs ?: return
        val edit = trustedPrefs.edit()
        val aliases = storageAliases(host, port)
        trustedPrefs.all.keys
            .filter { key -> aliases.any { alias -> key.startsWith("$alias|") } }
            .forEach(edit::remove)
        edit.apply()
    }

    private class TofuHostKeyRepository(
        private val prefs: SharedPreferences,
    ) : HostKeyRepository {
        override fun check(host: String?, key: ByteArray?): Int {
            val normalizedHost = host?.takeIf { it.isNotBlank() } ?: return HostKeyRepository.NOT_INCLUDED
            val publicKey = key ?: return HostKeyRepository.NOT_INCLUDED
            val hostKey = runCatching { HostKey(normalizedHost, publicKey) }.getOrNull()
                ?: return HostKeyRepository.NOT_INCLUDED
            val storageKey = storageKey(normalizedHost, hostKey.getType())
            val encoded = encode(publicKey)
            val rawStored = prefs.getString(storageKey, null)
            val trusted = readStored(rawStored)
            return when {
                trusted == null -> {
                    val handler = approvalRegistration?.handler
                    val approved = if (handler != null) {
                        val fp = runCatching {
                            val bytes = Base64.decode(encoded, Base64.NO_WRAP)
                            val md = java.security.MessageDigest.getInstance("SHA-256")
                            Base64.encodeToString(md.digest(bytes), Base64.NO_WRAP)
                                .trimEnd('=').let { "SHA256:$it" }
                        }.getOrDefault("(unknown)")
                        val deferred = CompletableDeferred<Boolean>()
                        awaitApproval(
                            handler,
                            HostKeyApprovalRequest(normalizedHost, hostKey.getType(), fp, deferred),
                        )
                    } else {
                        // No approval UI available (background worker / early init): fail closed.
                        // An unknown host must be trusted interactively before unattended use.
                        false
                    }
                    if (!approved) return HostKeyRepository.NOT_INCLUDED
                    persistApprovedFirstPin(
                        encoded = encoded,
                        readCurrent = { readStored(prefs.getString(storageKey, null)) },
                        write = { value ->
                            val toStore = SecretStore.encrypt(value) ?: value
                            prefs.edit().putString(storageKey, toStore).commit()
                        },
                    )
                }
                trusted == encoded -> {
                    // Key matches — migrate legacy plaintext entry to encrypted in-place
                    if (!SecretStore.isEncrypted(rawStored)) {
                        val toStore = SecretStore.encrypt(encoded) ?: encoded
                        prefs.edit().putString(storageKey, toStore).apply()
                    }
                    HostKeyRepository.OK
                }
                else -> HostKeyRepository.CHANGED
            }
        }

        override fun add(hostkey: HostKey?, ui: UserInfo?) {
            if (hostkey == null) return
            val encoded = hostkey.getKey()
            val toStore = SecretStore.encrypt(encoded) ?: encoded
            prefs.edit()
                .putString(storageKey(hostkey.getHost(), hostkey.getType()), toStore)
                .apply()
        }

        override fun remove(host: String?, type: String?) {
            if (host.isNullOrBlank()) return
            val edit = prefs.edit()
            matchingEntries(host, type).forEach { edit.remove(it.first) }
            edit.apply()
        }

        override fun remove(host: String?, type: String?, key: ByteArray?) {
            if (host.isNullOrBlank()) return
            val encodedKey = key?.let(::encode)
            val edit = prefs.edit()
            matchingEntries(host, type)
                .filter { encodedKey == null || it.second == encodedKey }
                .forEach { edit.remove(it.first) }
            edit.apply()
        }

        override fun getKnownHostsRepositoryID(): String = PREFS

        override fun getHostKey(): Array<HostKey> =
            prefs.all.mapNotNull { (key, value) -> toHostKey(key, readStored(value as? String)) }.toTypedArray()

        override fun getHostKey(host: String?, type: String?): Array<HostKey> =
            matchingEntries(host, type).mapNotNull { toHostKey(it.first, it.second) }.toTypedArray()

        private fun matchingEntries(host: String?, type: String?): List<Pair<String, String>> {
            val prefix = host?.takeIf { it.isNotBlank() }?.let { "${it}|" }
            return prefs.all.mapNotNull { (key, value) ->
                val decoded = readStored(value as? String) ?: return@mapNotNull null
                val parts = key.split("|", limit = 2)
                val entryHost = parts.getOrNull(0)
                val entryType = parts.getOrNull(1)
                val hostMatches = prefix == null || key.startsWith(prefix)
                val typeMatches = type.isNullOrBlank() || entryType == type
                if (entryHost != null && hostMatches && typeMatches) key to decoded else null
            }
        }

        private fun toHostKey(storageKey: String, encodedKey: String?): HostKey? {
            if (encodedKey.isNullOrBlank()) return null
            val parts = storageKey.split("|", limit = 2)
            val host = parts.getOrNull(0)?.takeIf { it.isNotBlank() } ?: return null
            return runCatching { HostKey(host, Base64.decode(encodedKey, Base64.NO_WRAP)) }.getOrNull()
        }

        // Reads a stored value: decrypt if encrypted, return as-is if legacy plaintext, null if blank.
        private fun readStored(raw: String?): String? {
            if (raw.isNullOrBlank()) return null
            return if (SecretStore.isEncrypted(raw)) SecretStore.decrypt(raw) else raw
        }

        private fun storageKey(host: String, type: String): String = "$host|$type"
        private fun encode(key: ByteArray): String = Base64.encodeToString(key, Base64.NO_WRAP)
    }

    /**
     * JSch calls HostKeyRepository synchronously, so bridge the UI decision with a strict deadline.
     * Completing an unanswered request as rejected also makes any late UI action a harmless no-op.
     */
    internal fun awaitApproval(
        handler: (HostKeyApprovalRequest) -> Unit,
        request: HostKeyApprovalRequest,
        timeoutMs: Long = APPROVAL_TIMEOUT_MS,
    ): Boolean {
        if (timeoutMs <= 0L) {
            request.deferred.complete(false)
            return false
        }
        return try {
            handler(request)
            runBlocking {
                withTimeoutOrNull(timeoutMs) { request.deferred.await() } ?: false
            }
        } catch (e: Exception) {
            if (e is InterruptedException) Thread.currentThread().interrupt()
            false
        } finally {
            // Fail closed on timeout, handler failure, cancellation, or thread interruption.
            request.deferred.complete(false)
        }
    }

    /**
     * Persist a TOFU approval without allowing two concurrent first connections to replace each
     * other's key. Both may have observed an empty preference before showing their dialogs, so the
     * pin must be re-read under one process-wide commit lock after approval. The first approved key
     * wins; an identical concurrent key is accepted, while a different key is reported as changed.
     */
    internal fun persistApprovedFirstPin(
        encoded: String,
        readCurrent: () -> String?,
        write: (String) -> Boolean,
    ): Int = synchronized(firstPinCommitLock) {
        when (readCurrent()) {
            null -> if (write(encoded)) HostKeyRepository.OK else HostKeyRepository.NOT_INCLUDED
            encoded -> HostKeyRepository.OK
            else -> HostKeyRepository.CHANGED
        }
    }
}

File: flutter_app/lib/data/ssh/ssh_host_key_trust.dart
/// Trust-on-first-use SSH host-key pinning, ported from `data/ssh/SshHostKeyTrust.kt`.
///
/// ## What changed, and why it is still equivalent
///
/// JSch handed its `HostKeyRepository` the **raw public key blob**, which the Kotlin stored base64
/// encoded. dartssh2's `SSHHostkeyVerifyHandler` is `(String type, Uint8List fingerprint)` — it
/// exposes only the OpenSSH-style SHA-256 fingerprint, and no host at all. So this port:
///
///  1. **Pins the fingerprint** (`SHA256:<base64, unpadded>`) rather than the key blob. That is not
///     a security downgrade: pinning a SHA-256 digest is as strong as pinning the preimage, and it
///     is exactly the value OpenSSH itself shows the user for comparison.
///  2. **Converts legacy pins on read.** An entry stored by the Kotlin app is a base64 public key;
///     [_canonicalFingerprint] derives its fingerprint using the same computation the Kotlin
///     `listKnownHosts` used, so an existing trust store keeps working and no host re-prompts. That
///     matters beyond convenience — re-prompting the whole fleet trains users to click through the
///     one dialog that is supposed to stop an interception.
///  3. **Builds the handler per connection**, closing over host and port, because dartssh2 does not
///     pass them.
///
/// The Kotlin needed `runBlocking` + `withTimeoutOrNull` because JSch called the repository
/// synchronously. dartssh2 awaits the handler, so the approval is a plain `Future` with a timeout.
library;

import 'dart:async';
import 'dart:convert';
import 'dart:typed_data';

import 'package:crypto/crypto.dart';

import '../remote_models.dart';
import 'async_lock.dart';


/// The outcome of checking a presented host key against the trust store.
enum HostKeyVerdict {
  /// Matches the pinned key: proceed.
  ok,

  /// A key is pinned for this host and it is **not** this one. Never auto-accepted.
  changed,

  /// Nothing pinned, and the connection was not approved (no UI, declined, or timed out).
  notIncluded,
}

/// A first-connect approval decision requested from the UI.
class HostKeyApprovalRequest {
  HostKeyApprovalRequest({
    required this.host,
    required this.keyType,
    required this.fingerprint,
    required this.completer,
  });

  final String host;
  final String keyType;

  /// OpenSSH-style `SHA256:…`, shown to the user to compare against the server's own output.
  final String fingerprint;
  final Completer<bool> completer;
}

/// Where pinned keys live. Abstracted so the trust logic is testable without a platform keystore.
abstract interface class HostKeyStore {
  Future<Map<String, String>> readAll();
  Future<String?> read(String key);
  Future<void> write(String key, String value);
  Future<void> delete(String key);
  Future<void> deleteAll();
}

/// In-memory store for tests.
class InMemoryHostKeyStore implements HostKeyStore {
  final Map<String, String> _entries = {};

  @override
  Future<Map<String, String>> readAll() async => Map.of(_entries);

  @override
  Future<String?> read(String key) async => _entries[key];

  @override
  Future<void> write(String key, String value) async => _entries[key] = value;

  @override
  Future<void> delete(String key) async => _entries.remove(key);

  @override
  Future<void> deleteAll() async => _entries.clear();
}

class SshHostKeyTrust {
  SshHostKeyTrust(this._store);

  final HostKeyStore _store;
  /// Serialises the read-then-write of a first pin.
  ///
  /// **Genuinely required, unlike the `@Synchronized` annotations elsewhere in the migration.**
  /// Those guarded purely synchronous methods, which cannot interleave on a single-threaded isolate.
  /// Here the store read is `await`ed, so a second connection to the same host can run between this
  /// one's read and its write — exactly the race the Kotlin `synchronized(firstPinCommitLock)`
  /// block existed to close.
  final AsyncLock _firstPinLock = AsyncLock();

  static const approvalTimeout = Duration(seconds: 120);

  Object? _approvalOwner;
  void Function(HostKeyApprovalRequest)? _approvalHandler;

  /// Registration is owner-token based: a retiring ViewModel may clear only its own handler, never a
  /// newer instance's.
  void registerApprovalHandler(Object owner, void Function(HostKeyApprovalRequest) handler) {
    _approvalOwner = owner;
    _approvalHandler = handler;
  }

  void clearApprovalHandler(Object owner) {
    if (identical(_approvalOwner, owner)) {
      _approvalOwner = null;
      _approvalHandler = null;
    }
  }

  // ── storage-key conventions ────────────────────────────────────────────────

  /// Every storage-key alias prefix a given host/port can be pinned under.
  ///
  /// JSch wrote bare `host` for port 22 and `[host]:port` otherwise, and older builds also produced
  /// `host:port`. All three are still recognised so an existing trust store keeps working.
  static Set<String> storageAliases(String host, int port) => {
        host,
        '$host:$port',
        '[$host]:$port',
      };

  /// The alias new pins are written under, matching JSch's own convention.
  static String canonicalAlias(String host, int port) => port == 22 ? host : '[$host]:$port';

  static String _storageKey(String alias, String keyType) => '$alias|$keyType';

  /// Normalises a stored value to an OpenSSH-style fingerprint.
  ///
  /// Values written by this app are already `SHA256:…`. Values inherited from the Kotlin app are a
  /// base64 public key, and are hashed here with the same computation `listKnownHosts` used, so a
  /// migrated trust store verifies without re-prompting.
  static String? _canonicalFingerprint(String? stored) {
    if (stored == null || stored.trim().isEmpty) return null;
    final value = stored.trim();
    if (value.startsWith('SHA256:')) return value;
    try {
      final digest = sha256.convert(base64.decode(value));
      return 'SHA256:${base64.encode(digest.bytes).replaceAll('=', '')}';
    } catch (_) {
      // Unparseable entry: treat as absent rather than as a match. Failing closed here means the
      // user is asked to approve again, never that a corrupt row silently authorises a host.
      return null;
    }
  }

  /// The fingerprint string for a raw public key, matching what dartssh2 reports.
  static String fingerprintOfKey(Uint8List publicKey) {
    final digest = sha256.convert(publicKey);
    return 'SHA256:${base64.encode(digest.bytes).replaceAll('=', '')}';
  }

  /// Decodes the fingerprint dartssh2 passes to its verify handler, which is the UTF-8 bytes of the
  /// `SHA256:<base64>` string rather than the digest itself.
  static String decodeHandlerFingerprint(Uint8List fingerprint) => utf8.decode(fingerprint);

  // ── the trust check ────────────────────────────────────────────────────────

  /// Checks a presented key, prompting for approval when nothing is pinned yet.
  Future<HostKeyVerdict> check({
    required String host,
    required int port,
    required String keyType,
    required String fingerprint,
  }) async {
    final aliases = storageAliases(host, port);
    final all = await _store.readAll();

    // A pin under any alias counts: the same host reached as "nas" and "[nas]:2222" is one host.
    String? pinned;
    for (final entry in all.entries) {
      final parts = entry.key.split('|');
      if (parts.length < 2) continue;
      final entryType = entry.key.substring(parts[0].length + 1);
      if (!aliases.contains(parts[0]) || entryType != keyType) continue;
      final canonical = _canonicalFingerprint(entry.value);
      if (canonical != null) {
        pinned = canonical;
        break;
      }
    }

    if (pinned != null) return pinned == fingerprint ? HostKeyVerdict.ok : HostKeyVerdict.changed;

    final handler = _approvalHandler;
    if (handler == null) {
      // No approval UI available (background worker / early init): fail closed. An unknown host
      // must be trusted interactively before unattended use.
      return HostKeyVerdict.notIncluded;
    }

    final approved = await awaitApproval(
      handler,
      HostKeyApprovalRequest(
        host: host,
        keyType: keyType,
        fingerprint: fingerprint,
        completer: Completer<bool>(),
      ),
    );
    if (!approved) return HostKeyVerdict.notIncluded;

    return persistApprovedFirstPin(
      storageKey: _storageKey(canonicalAlias(host, port), keyType),
      fingerprint: fingerprint,
    );
  }

  /// Bridges the UI decision with a strict deadline.
  ///
  /// Fails closed on timeout, handler failure or a completer that is never answered, so an
  /// unanswered prompt can never leave a connection authorised. Completing the request as rejected
  /// also makes any late UI action a harmless no-op.
  Future<bool> awaitApproval(
    void Function(HostKeyApprovalRequest) handler,
    HostKeyApprovalRequest request, {
    Duration timeout = approvalTimeout,
  }) async {
    if (timeout <= Duration.zero) {
      if (!request.completer.isCompleted) request.completer.complete(false);
      return false;
    }
    try {
      handler(request);
      return await request.completer.future.timeout(timeout, onTimeout: () => false);
    } catch (_) {
      return false;
    } finally {
      if (!request.completer.isCompleted) request.completer.complete(false);
    }
  }

  /// Persists a TOFU approval without letting two concurrent first connections replace each other's
  /// key.
  ///
  /// Both may have observed an empty store before showing their dialogs, so the pin is re-read under
  /// the commit lock after approval. The first approved key wins; an identical concurrent key is
  /// accepted, while a different key is reported as changed.
  Future<HostKeyVerdict> persistApprovedFirstPin({
    required String storageKey,
    required String fingerprint,
  }) =>
      _firstPinLock.synchronized(() async {
        final current = _canonicalFingerprint(await _store.read(storageKey));
        if (current == null) {
          await _store.write(storageKey, fingerprint);
          return HostKeyVerdict.ok;
        }
        return current == fingerprint ? HostKeyVerdict.ok : HostKeyVerdict.changed;
      });

  // ── trust-store management ─────────────────────────────────────────────────

  Future<List<KnownHost>> listKnownHosts() async {
    final all = await _store.readAll();
    final out = <KnownHost>[];
    all.forEach((storageKey, rawValue) {
      final fingerprint = _canonicalFingerprint(rawValue);
      if (fingerprint == null) return;
      final sep = storageKey.indexOf('|');
      if (sep <= 0) return;
      out.add(KnownHost(
        storageKey.substring(0, sep),
        storageKey.substring(sep + 1),
        fingerprint,
      ));
    });
    return out;
  }

  /// True when at least one host key (any type) is already pinned for this host.
  Future<bool> hasPinnedKey(String host, int port) async {
    final aliases = storageAliases(host, port);
    final all = await _store.readAll();
    return all.keys.any((key) => aliases.any((alias) => key.startsWith('$alias|')));
  }

  Future<void> removeHost(String host, int port) async {
    final aliases = storageAliases(host, port);
    final all = await _store.readAll();
    for (final key in all.keys) {
      if (aliases.any((alias) => key.startsWith('$alias|'))) {
        await _store.delete(key);
      }
    }
  }

  /// All pinned host keys as storage key (`host|type`) → fingerprint, used by the app backup so a
  /// restored fleet keeps its trust store.
  Future<Map<String, String>> exportEntries() async {
    final all = await _store.readAll();
    final out = <String, String>{};
    all.forEach((key, value) {
      final fingerprint = _canonicalFingerprint(value);
      if (fingerprint != null) out[key] = fingerprint;
    });
    return out;
  }

  /// Import pinned host keys from a backup.
  ///
  /// **Existing pins always win** — a backup must never silently replace a key this device has
  /// already verified, which would turn a restore into an interception vector.
  Future<void> importEntries(Map<String, String> entries) async {
    final existing = await _store.readAll();
    for (final entry in entries.entries) {
      if (!entry.key.contains('|') || entry.value.trim().isEmpty) continue;
      if (existing.containsKey(entry.key)) continue;
      final fingerprint = _canonicalFingerprint(entry.value);
      if (fingerprint != null) await _store.write(entry.key, fingerprint);
    }
  }

  /// Replace the complete trust snapshot, used only to compensate a failed transactional restore.
  Future<void> replaceEntries(Map<String, String> entries) async {
    await _store.deleteAll();
    for (final entry in entries.entries) {
      if (!entry.key.contains('|') || entry.value.trim().isEmpty) continue;
      final fingerprint = _canonicalFingerprint(entry.value);
      if (fingerprint != null) await _store.write(entry.key, fingerprint);
    }
  }

  /// Filter a backup's exported trust entries down to those whose host belongs to one of [hosts].
  ///
  /// Used at restore time so pinned keys are imported only for servers actually restored — orphaned
  /// trust entries for skipped/limited hosts are dropped.
  static Map<String, String> filterEntriesForHosts(
    Map<String, String> entries,
    Iterable<(String host, int port)> hosts,
  ) {
    if (entries.isEmpty) return entries;
    final aliases = <String>{for (final (host, port) in hosts) ...storageAliases(host, port)};
    return {
      for (final entry in entries.entries)
        if (aliases.any((alias) => entry.key.startsWith('$alias|'))) entry.key: entry.value,
    };
  }
}
